Make the NewParallel scheduler the default, even for -j1 builds

acmorrow opened the pull request:
- Removes support for `--experimental=tm_v2` and makes it the default
- Adds `--experimental=legacy_sched` to opt-in to the old behavior
- Fixes a bug in `SConf.py` where the cached exception was not raised, and the code assumed the current thread had an 'active' exception.
- Uses `NewParallel` for `-j1` builds as well, but disables locking
- Changes `NewParallel` to use the main thread as a worker for `-j1` builds, bringing it more into alignment with the single threaded nature of `Serial`. Parallel builds still block the main thread and use only worker threads, so that all working threads are on the same footing w.r.t. any stack size settings.
- Fixes a broken f-string in `self.trace_message("Found {len(results_queue)} completed tasks to process")`
- Fixes *most* tests that I was able to run locally. The `test/Interactive/taskmastertrace.py` and `test/option/taskmastertrace.py` are still broken since the new scheduler has additional task master tracing output and breaks the comparisons, and I'm not quite sure how to fix it.

Note that I've only run the test suite on my mac and I don't have all the tools available (e.g. Java). It is possible that there are further regressions that I haven't seen, but hopefully the CI run here will reveal that. I'm making this PR draft until decisions are made about any failing tests.

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

acmorrow commented:
@bdbaddog - Note that the indentation changes and some other things make this a challenge to review, because there are also changes to the actual classes stacked above it. Let me know if you want me to try to make just one PR to pull out the "do we have threading" stuff and then stack this above that.

acmorrow commented:
Another thing to note: we now assume that the `threading` package is available. This was true as of CPython 3.7: https://docs.python.org/3.8/whatsnew/3.7.html#build-changes and https://github.com/python/cpython/issues/75551.

Do we care much about any other pythons which could in theory lack the `threading` package?

bdbaddog commented on CHANGES.txt:
```diff
@@ -61,6 +61,11 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
       expected bytecodes in ActionTests.py.
 
 
+  From Andrew Morrow:
```
Move this up a bit, should be alphabetical by last name.

acmorrow replied:
Done

bdbaddog commented on SCons/Script/SConsOptions.py:
```diff
@@ -40,7 +40,7 @@
 diskcheck_all = SCons.Node.FS.diskcheck_types()
 
-experimental_features = {'warp_speed', 'transporter', 'ninja', 'tm_v2'}
+experimental_features = {'warp_speed', 'transporter', 'ninja', 'legacysched'}
```
how about legacy_sched (ala warp_speed).. easier to read/type correctly

acmorrow replied:
Done

acmorrow commented on SCons/Taskmaster/Job.py:
```diff
@@ -239,505 +228,523 @@ def start(self):
         self.taskmaster.cleanup()
 
 
-# Trap import failure so that everything in the Job module but the
-# Parallel class (and its dependent classes) will work if the interpreter
-# doesn't support threads.
-try:
-    import queue
-    import threading
-except ImportError:
-    pass
-else:
-    class Worker(threading.Thread):
```
@bdbaddog - Let me know if you want a PR that is *just* the removal of this check and then another that is just the de-indent. Then I could try to rebase this PR above those after they merge, and maybe the real diffs will be a lot easier to see.

acmorrow replied:
This might help avoid the need for that. Here is an "ignore whitespace" view of the very first commit in the stack, where I did the removal and de-indent: https://github.com/acmorrow/scons/commit/8c8052c95d4645a584b4332e11aabe0843446678?diff=split&w=1. That way you can see what really changed there in the various schedulers is, effectively, nothing.

Then, review the PR without that bottom commit, like: https://github.com/SCons/scons/pull/4464/files/8c8052c95d4645a584b4332e11aabe0843446678..e4e128a70ea7659b95c5edcc713436683d9a2cca

bdbaddog replied:
This file is fine as is.

mwichmann commented:
> Another thing to note: we now assume that the `threading` package is available. This was true as of CPython 3.7: https://docs.python.org/3.8/whatsnew/3.7.html#build-changes and [python/cpython#75551](https://github.com/python/cpython/issues/75551).
> 
> Do we care much about any other pythons which could in theory lack the `threading` package?

Seems unlikely. @bdbaddog has suggested dropping 3.6 a few times, and I've offered the opinion we don't really need to rush to do that, mainly because of RHEL/CentOS/Rocky/Alma 8, which still live. But the Python there definitely is a threaded build.

acmorrow commented:
> The test/Interactive/taskmastertrace.py and test/option/taskmastertrace.py are still broken since the new scheduler has additional task master tracing output and breaks the comparisons, and I'm not quite sure how to fix it.

@bdbaddog - Any suggestions on the best way to proceed here? Part of the issue is that the new taskmaster output has varying data in it, and the other is that places where taskmaster tracing output only included node transitions now includes scheduler state transition notes.

bdbaddog commented:
@acmorrow  for `test/option/taskmastertrace.py` there are "gold" files in `test/option/fixtures` which that test uses.
Probably worth keeping the old golds for legacy_sched to use, and create new ones in that fixture dir.

Most of those are no -j specified, but one is -j2. Is the new scheduler stable for -j2 ?

Also, maybe copy this test for taskmastertrace-legacy.py and keep the existing golden files, until/if we remove the old scheduler.

bdbaddog commented:
@acmorrow for `test/Interactive/taskmastertrace.py` looks like similar, there's golden output. I'd say if you're willing move that text to fixture golden file, and add for --experimental=legacy_sched as well.

bdbaddog commented on SCons/SConf.py:
```diff
@@ -253,8 +253,7 @@ def failed(self):
         if issubclass(exc_type, SConfError):
-            # TODO pylint E0704: bare raise not inside except
-            raise
+            raise self.exc_info()[1]
```
can you add a comment what this is?

acmorrow replied:
I will if you think it is necessary but I sort of think the line up on 254 and the comment above it make it reasonably clear already. Actually clearer now because now both cases are acting on the state in `self.exc_info()`.

acmorrow replied:
@bdbaddog - What do you think? Does it need more commentary?

mwichmann replied:
Style note: normally, I'd grumble about calling a function/method twice right by each other, although in this case the `exc_info` method is a trivial one-liner returning an attribute. Seems like you could, nonetheless, unpack the exception tuple in the first call instead of having to use `[0]` and `[1]` just for readability.

bdbaddog replied:
@acmorrow - nope. you're right 253's comment is sufficient

bdbaddog commented on SCons/Taskmaster/JobTests.py:
```diff
@@ -284,50 +285,6 @@ def runTest(self):
         self.assertFalse(taskmaster.num_failed,
                     "some task(s) failed to execute")
 
-        # Verify that parallel jobs will pull all of the completed tasks
```
Not sure we should delete this quite yet if we're still providing legacy_sched?
@mwichmann  thoughts?

acmorrow replied:
Yeah, I was a little fed up with this test and maybe overreacted a bit and just deleted it. But every time I've experimented with scheduler stuff it got in the way. Is there a way to temporarily opt-in to `--experimental=legacy_sched` here which would make the test pass?

acmorrow replied:
@bdbaddog - What do you think? Restore the test but force it to `legacy_sched` somehow? Or do away with it. It really is testing a behavior of the legacy scheduler, so I don't see that there is much value in adapting it for the new scheduler.

bdbaddog replied:
Yes. restore. There's got to be some prior art in the test suite about forcing a GetOption(), maybe calling SetOption() before in the test before the rest is run?

acmorrow replied:
I figured out how to do it, quite easy. Restored.

acmorrow commented:
> > Another thing to note: we now assume that the `threading` package is available. This was true as of CPython 3.7: https://docs.python.org/3.8/whatsnew/3.7.html#build-changes and [python/cpython#75551](https://github.com/python/cpython/issues/75551).
> > Do we care much about any other pythons which could in theory lack the `threading` package?
> 
> Seems unlikely. @bdbaddog has suggested dropping 3.6 a few times, and I've offered the opinion we don't really need to rush to do that, mainly because of RHEL/CentOS/Rocky/Alma 8, which still live. But the Python there definitely is a threaded build.

@bdbaddog - What do you say? Should we bump min python to 3.7 in this review for guaranteed `threading`, or keep at 3.6 for RHEL8 and associates despite the slight risk of encountering a non-threaded 3.6?

acmorrow commented on test/option/fixture/taskmaster_expected_new_parallel.txt:
```diff
@@ -1,6 +1,6 @@
 Job.NewParallel._work(): [Thread:XXXXX] Gained exclusive access
 Job.NewParallel._work(): [Thread:XXXXX] Starting search
-Job.NewParallel._work(): [Thread:XXXXX] Found {len(results_queue)} completed tasks to process
+Job.NewParallel._work(): [Thread:XXXXX] Found 0 completed tasks to process
```
This is sufficient to get `test/option/taskmastertrace.py` to pass with the fixed f-string in Job.py.

acmorrow commented on test/option/taskmastertrace.py:
```diff
@@ -42,7 +42,7 @@
 expect_stdout = test.wrap_stdout(test.read('taskmaster_expected_stdout_1.txt', mode='r'))
 
-test.run(arguments='--taskmastertrace=- .', stdout=expect_stdout)
+test.run(arguments='--experimental=legacy_sched --taskmastertrace=- .', stdout=expect_stdout)
```
I'd just missed this one somehow; the test now passes.

acmorrow commented:
@bdbaddog - The `macos-latest` build seems also to be failing on master so I'm disregarding those failures.

bdbaddog commented:
> > > Another thing to note: we now assume that the `threading` package is available. This was true as of CPython 3.7: https://docs.python.org/3.8/whatsnew/3.7.html#build-changes and [python/cpython#75551](https://github.com/python/cpython/issues/75551).
> > > Do we care much about any other pythons which could in theory lack the `threading` package?
> > 
> > 
> > Seems unlikely. @bdbaddog has suggested dropping 3.6 a few times, and I've offered the opinion we don't really need to rush to do that, mainly because of RHEL/CentOS/Rocky/Alma 8, which still live. But the Python there definitely is a threaded build.
> 
> @bdbaddog - What do you say? Should we bump min python to 3.7 in this review for guaranteed `threading`, or keep at 3.6 for RHEL8 and associates despite the slight risk of encountering a non-threaded 3.6?

Yes. And add a TODO to remove when/if we drop the legacy scheduler so it's easy to find..

acmorrow commented:
@mwichmann and @bdbaddog - OK, updated, PTAL. All tests passing, comments addressed and resolved. The only outstanding item as far as I know is this:
```
@bdbaddog - What do you say? Should we bump min python to 3.7 in this review for guaranteed threading, or keep at 3.6 for RHEL8 and associates despite the slight risk of encountering a non-threaded 3.6?

Yes. And add a TODO to remove when/if we drop the legacy scheduler so it's easy to find..
```

I'm not sure exactly what you want me to do though. Can you be annoyingly specific?

acmorrow commented:
@bdbaddog and @mwichmann - I think I've addressed all comments and requests.

- Here is the lone commit that removed the `threading` check from `Job.py`: https://github.com/SCons/scons/pull/4464/commits/8c8052c95d4645a584b4332e11aabe0843446678. But git(hub?) does a terrible job seeing that its almost all just indent changes, here is the same commit without whitespace: https://github.com/SCons/scons/pull/4464/commits/8c8052c95d4645a584b4332e11aabe0843446678?diff=split&w=1
- Here is the remainder of the changes *without* the above change: https://github.com/SCons/scons/pull/4464/files/8c8052c95d4645a584b4332e11aabe0843446678..36b6ab73bd3597694d4edf9117d67f826280ed40

Please let me know if I've missed anything as there has been a lot of discussion around this, but hopefully this is ready to squash and merge.

bdbaddog commented:
Thanks!
